CLR_Retraining.py

```diff
@@ -0,0 +1,144 @@
+
+# Load and preprocess the example image
+url_image, filename_image = ("https://github.com/pytorch/hub/raw/master/images/dog.jpg", "dog.jpg")
+urllib.request.urlretrieve(url_image, filename_image)
+
+input_image = Image.open(filename_image)
+preprocess = transforms.Compose([
+    transforms.Resize(256),
+    transforms.CenterCrop(224),
+    transforms.ToTensor(),
+    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
+])
+input_tensor = preprocess(input_image)
+input_batch = input_tensor.unsqueeze(0)
+
+# Create the validation_loader
+validation_set = datasets.ImageFolder('/Users/emiliamosthaf/Documents/viertes semester/teamprojekt/imagenette2-320',
+                                      transform=transforms.Compose([
+                                          transforms.Resize(256),
+                                          transforms.CenterCrop(224),
+                                          transforms.ToTensor(),
+                                          transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                                               std=[0.229, 0.224, 0.225]),
+                                      ]))
+
+# Use a smaller subset for validation
+validation_indices = range(100)
+validation_sampler = SubsetRandomSampler(validation_indices)
+validation_loader = DataLoader(validation_set, batch_size=128, sampler=validation_sampler, num_workers=0)
+
+# Create the training_loader
+train_set = datasets.ImageFolder('/Users/emiliamosthaf/Documents/viertes semester/teamprojekt/imagenette2-320',
+                                 transform=transforms.Compose([
+                                     transforms.Resize(256),
+                                     transforms.CenterCrop(224),
+                                     transforms.ToTensor(),
+                                     transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                                          std=[0.229, 0.224, 0.225]),
+                                 ]))
+train_loader = DataLoader(train_set, batch_size=128,sampler=validation_sampler , num_workers=0)
+
+# Check if train_loader has data
+if len(train_loader) == 0:
+    raise ValueError("The training data loader is empty. Please check your training dataset path and contents.")
+
+# Loss function and optimizer
+criterion = nn.CrossEntropyLoss()
+
+# Pruning amount for the loop
+amounts = [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
+
+# Create a multidimensional array that can store all the values
+results_global_unstructured_l1 = np.zeros(len(amounts))
+
+# Function to apply CLR
+def train_model_with_clr(model, train_loader, criterion, base_lr, max_lr, steps_per_epoch, epochs):
+    optimizer = optim.SGD(model.parameters(), lr=base_lr, momentum=0.9)
+    scheduler = torch.optim.lr_scheduler.CyclicLR(optimizer, base_lr=base_lr, max_lr=max_lr, step_size_up=max(1, steps_per_epoch//2), mode='triangular')
+    
```

Comment:
just for informational purposes, maybe printing the current lr for each epoch would be interesting to check if the cycle from base to max and back to base works as it should